---
layout: default
---
<script src="vendor/handsontable-master/dist/handsontable.full.min.js"></script>
<link rel="stylesheet" href="vendor/handsontable-master/dist/handsontable.full.min.css">


<div class="container-fluid">

  <div class="row">
    <div class="col-sm-3">
      <div class="panel panel-primary">
	<div class="panel-heading">
	  <h3 class="panel-title pull-left">About UBiT&sup2;</h3>
	  <div class="btn-group pull-right">
	    <a class="btn btn-primary btn-sm" href="about.html">
	      <i class="fa fa-lg fa-info-circle"></i>
	    </a>
	    <a class="btn btn-primary btn-sm" data-toggle="collapse" href="#infoClose" aria-expanded="false" aria-controls="infoClose">
	      <i class="fa fa-lg fa-minus-square-o"></i>
	    </a>
	  </div>
	  <div class="clearfix"></div>
	</div>
	<div class="panel-body collapse in" id="infoClose">
	  UBiT&sup2; enables fast, simple bioinformatics analysis online. Your data is never accessed or stored in any way. 
	</div>
      </div>
    </div>
    
    <!--- data panel --->
    <div class="col-sm-9">
      <div class="panel panel-default">
	<div class="panel-heading">
	  <h3 class="panel-title pull-left">Data</h3>
	  <div class="btn-group pull-right">
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#dataDesc" aria-expanded="false" aria-controls="dataDesc">
	      <i class="fa fa-lg fa-question"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#dataOptions" aria-expanded="false" aria-controls="dataOptions">
	      <i class="fa fa-gear"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#dataClose" aria-expanded="false" aria-controls="dataClose">
	      <i class="fa fa-lg fa-minus-square-o"></i>
	    </a>
	  </div>
	  <div class="clearfix"></div>
	  <div class="collapse" id="dataDesc">
	    <div class="card card-block">
	      Gene expression matrix with rows as samples and columns as genes. Sample comma delimited Ct values for BioMark data from <a href="http://www.nature.com/nature/journal/v526/n7571/full/nature15260.html"><i>Lawson et al.</i></a> are provided as an example. LoD transformation assumes failures as 999 and a LoD Ct value of 28 such that Log2 expression can be approximated as 28 - Ct values. Use the options to specify different delimiters as needed. Check the Transpose option if your data is rows as genes and columns as samples.
	    </div>
	  </div>
	  <div class="collapse" id="dataOptions">
	    Transform data:
	    <select id="transform" class="btn btn-default">
	      <option value="none">none</option>
	      <option value="biomark" selected>LoD method (rt-qpcr/Biomark)</option>
	      <option value="log10">log10</option>
	      <option value="asinh">asinh</option>
	    </select>
	    Delimiter:
	    <select id="delimiter" class="btn btn-default">
	      <option value="comma" selected>, (comma)</option>
	      <option value="tab">\t (tab)</option>
	    </select>
	    Transpose?: <input type="checkbox" id="transpose">
	  </div>	    
	</div>
	<div class="panel-body collapse in" id="dataClose">
	  <div style="height:100px; overflow: scroll">
	    <div id="incsv" class="hot handsontable"></div>
	  </div>
	  <br>
	  <button type="button" id="inbut" class="btn btn-primary btn-block">analyze</button>
	</div>
      </div>
    </div>      
  </div>
  
  <div class="row">      
    <div class="col-sm-6">
      <div class="panel panel-default">
	<div class="panel-heading">
	  <h3 class="panel-title pull-left">% Success Per Gene</h3>
	  <div class="btn-group pull-right">
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#percentage_success_per_gene_desc" aria-expanded="false" aria-controls="percentage_success_per_gene_desc">
	      <i class="fa fa-lg fa-question"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#percentage_success_per_gene_options" aria-expanded="false" aria-controls="percentage_success_per_gene_options">
	      <i class="fa fa-gear"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#percentage_success_gene_close" aria-expanded="false" aria-controls="percentage_success_gene_close">
	      <i class="fa fa-lg fa-minus-square-o"></i>
	    </a>
	  </div>
	  <div class="clearfix"></div>
	  <div class="collapse" id="percentage_success_per_gene_desc">
	    <div class="card card-block">
	      Histogram summarizing the distribution of and barplot visualizing the percentage of successes (non-failures) per gene. Failed genes identified as having fewer percentage successes than a user-defined threshold, highlighted in red, are removed from downstream analysis. Use the threshold slider to adjust the threshold.
	    </div>
	  </div>	  
	  <div class="collapse" id="percentage_success_per_gene_options">
	    Percentage Threshold:
	    <input type="text" value="30" data-slider-min="0" data-slider-max="100" data-slider-step="10" id="percentage_success_per_gene_threshold_slider" data-slider-orientation="horizontal" data-slider-selection="after" data-slider-tooltip="show">
	    <script>$('#percentage_success_per_gene_threshold_slider').slider({
	      formater: function(value) {
	      return 'Current value: '+value;
	      }
	      });</script>
	    <button type="button" id="rerun_percentage_success_per_gene" class="btn btn-primary btn-block">reset threshold</button>
	  </div>	    
	</div>
	<div class="panel-body collapse in" id="percentage_success_gene_close">  
	  <div style="height:150px" id="hist_percentage_success_per_gene">
	    <div id="hist1"></div>
	  </div>
	  <br/>
	  <div style="height:150px" id="barplot_percentage_success_per_gene">
	    <div id="hist1"></div>
	  </div>
	</div>
      </div>	
    </div>
      
    <div class="col-sm-6">
      <div class="panel panel-default">
	<div class="panel-heading">
	  <h3 class="panel-title pull-left">% Genes Detected Per Sample</h3>
	  <div class="btn-group pull-right">
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#percentage_genes_detected_per_sample_desc" aria-expanded="false" aria-controls="percentage_genes_detected_per_sample_desc">
	      <i class="fa fa-lg fa-question"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#percentage_genes_detected_per_sample_options" aria-expanded="false" aria-controls="percentage_genes_detected_per_sample_options">
	      <i class="fa fa-gear"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#percentage_genes_detected_per_sample_close" aria-expanded="false" aria-controls="percentage_gene_detected_per_sample_close">
	      <i class="fa fa-lg fa-minus-square-o"></i>
	    </a>
	  </div>
	  <div class="clearfix"></div>
	  <div class="collapse" id="percentage_genes_detected_per_sample_desc">
	    <div class="card card-block">
	      Histogram summarizing the distribution of and barplot visualizing the percentage of genes detected per sample. Failed samples identified as having fewer percentage genes detected than a user-defined threshold, highlighted in red, are removed from downstream analysis. Use the threshold slider to adjust the threshold.
	    </div>
	  </div>	  
	  <div class="collapse" id="percentage_genes_detected_per_sample_options">
	      Percentage Threshold:
	      <input type="text" value="30" data-slider-min="0" data-slider-max="100" data-slider-step="10" id="percentage_genes_detected_per_sample_slider" data-slider-orientation="horizontal" data-slider-selection="after" data-slider-tooltip="show">
	      <script>$('#percentage_genes_detected_per_sample_slider').slider({
		formater: function(value) {
		return 'Current value: '+value;
		}
		});</script>
	      <button type="button" id="rerun_percentage_genes_detected_per_sample" class="btn btn-primary btn-block">reset threshold</button>
	  </div>	    	    
	</div>
	<div class="panel-body collapse in" id="percentage_genes_detected_per_sample_close">  
	  <div style="height:150px" id="hist_percentage_genes_detected_per_sample">
	    <div id="hist2"></div>
	  </div>
	  <br>
	  <div style="height:150px" id="barplot_percentage_genes_detected_per_sample">
	    <div id="hist2"></div>
	  </div>
	</div>
      </div>
    </div>     
  </div>
  
  <!--- main analysis --->
  <div class="row">         
    <!--- hcluster panel --->
    <div class="col-sm-8">
      <div class="panel panel-default">
	<div class="panel-heading">
	  <h3 class="panel-title pull-left">Hierarchical Clustering Heatmap</h3>
	  <div class="btn-group pull-right">
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#hclustDesc">
	      <i class="fa fa-lg fa-question"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#hclustOptions" aria-expanded="false" aria-controls="hclustOptions">
	      <i class="fa fa-gear"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#hclustClose" aria-expanded="false" aria-controls="hclustClose">
	      <i class="fa fa-lg fa-minus-square-o"></i>
	    </a>
	  </div>
	  <div class="clearfix"></div>
	  <div class="collapse" id="hclustDesc">
	    <div class="card card-block">
	      Unbiased row-wise (sample) hierarchical clustering with heatmap visualization. Heatmap colors are column-scaled. Use options to specify the distance metric and linkage criterion. Columns (genes) are ordered in the same order as in the Differential Expression panel. 
	    </div>
	  </div>	  
	  <div class="collapse" id="hclustOptions">
	    Linkage criterion:
	    <select id="cluster_method" class="btn btn-default">
	      <option value="complete" selected>complete</option>
	      <option value="average">average</option>
	      <option value="single">single</option>
	    </select>
	    Distance metric:
	    <select id="distance_metric" class="btn btn-default">
	      <option value="euclidean" selected>euclidean</option>
	      <option value="corr">correlation</option>
		<option value="manhattan">manhattan</option>
	      <option value="max">max</option>
	    </select>
	    <button type="button" id="rerunHclust" class="btn btn-primary btn-block">rerun hierarchical clustering</button>
	  </div>	    
	</div>	  
	<div class="panel-body collapse in" id="hclustClose">
	  <div style="height:363px" id="heatmap_panel">	    
	    <div id="dendro"></div>
	  </div>
	</div>
	
      </div>
    </div>
    
    <div class="col-sm-4">
      <div class="panel panel-default">
	<div class="panel-heading">
	  <h3 class="panel-title pull-left">Principal Component Analysis</h3>
	  <div class="btn-group pull-right">
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#pcaDesc">
	      <i class="fa fa-lg fa-question"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#pcaOptions" aria-expanded="false" aria-controls="pcaOptions">
	      <i class="fa fa-gear"></i>
	    </a>
	    <a class="btn btn-default btn-sm" data-toggle="collapse" href="#pcaClose" aria-expanded="false" aria-controls="pcaClose">
	      <i class="fa fa-lg fa-minus-square-o"></i>
	    </a>
	    </div>
	  <div class="clearfix"></div>
	  <div class="collapse" id="pcaDesc">
	    <div class="card card-block">
	      Principal component analysis with K-means clustering with K = 2 to identify the main 2 subpopulations. For principal component analysis, to ensure an invertible matrix, more rows than columns are needed. Data will be trimmed to the necessary dimensions for PCA as needed. Differential expression and other downstream analyses are calculated according to these subpopulation group labels.
	    </div>
	  </div>	  
	  <div class="collapse" id="pcaOptions">
	    <button type="button" id="rerunKmeans" class="btn btn-primary btn-block">rerun K-means clustering</button>
	  </div>
	</div>
	<div class="panel-body collapse in" id="pcaClose">
	  <div style="height:363px" id="pca_panel">
	    <div id="pca"></div>
	  </div>
	</div>
      </div>
    </div>
  </div>

  <!--- analysis row 2 --->
  <div class="row">
    
      <div class="col-sm-8">
	<div class="panel panel-default">
	  <div class="panel-heading">
	    <h3 class="panel-title pull-left">Differential Expression</h3>
	    <div class="btn-group pull-right">
	      <a class="btn btn-default btn-sm" data-toggle="collapse" href="#diffexpDesc">
		<i class="fa fa-lg fa-question"></i>
	      </a>
	      <a class="btn btn-default btn-sm" data-toggle="collapse" href="#diffexpOptions" aria-expanded="false" aria-controls="diffexpOptions">
		<i class="fa fa-gear"></i>
	      </a>
	      <a class="btn btn-default btn-sm" data-toggle="collapse" href="#diffexpClose">
		<i class="fa fa-lg fa-minus-square-o"></i>
	      </a>
	    </div>
	    <div class="clearfix"></div>
	    <div class="collapse" id="diffexpDesc">
	      <div class="card card-block">
		Barplot showing distribution of p-values with significant hits (p-value < 0.05 / # of tests) highlighted in red. P-values calculated by Wilcoxon rank-sum test. Likewise, the barplot of fold-changes is also plotted. A volcano plot shows p-values vs. fold-change with significant hits highlighted in red. Use options to sort by fold-change or p-value and update barplots. Sorting influenced column-ordering in the Heatmap map and the ranking used in the Gene Set Enrichment panel.
	      </div>
	    </div>	  
	    <div class="collapse" id="diffexpOptions">
	      Sort by:
	      <select id="diffexp_sort" class="btn btn-default">
		<option value="fold-change" selected>fold-change</option>
		<option value="p-value">p-value</option>
	      </select>
	      <select id="diffexp_sortdir" class="btn btn-default">
		<option value="decreasing" selected>decreasing</option>
		<option value="increasing">increasing</option>
	      </select>
	      <button type="button" id="rerunDiffexp" class="btn btn-primary btn-block">rerun differential expression analysis</button>
	    </div>	    
	  </div>
	  <div class="panel-body collapse in" id="diffexpClose">
	    <div class="row">
	      <div class="col-sm-6">
		<div style="height:336px" id="volcano_panel">
		  <b>Volcano Plot</b>
		  <div id="volcano"></div>
		</div>
	      </div>
	      <div class="col-sm-6">
		<div style="height:150px" id="diffexp_pval_panel">
		  <b>Distribution of -log10(P-value)</b>
		  <div id="diffexp_pval"></div>
		</div>
		<br>
		<div style="height:150px" id="diffexp_fc_panel">
		  <b>Distribution of log2(fold change)</b>
  		  <div id="diffexp_fc"></div>
		</div>
	      </div>
	    </div>
          </div>
	</div>

      </div>

      <div class="col-sm-4">
	<div class="panel panel-default">
	  <div class="panel-heading">
	    <h3 class="panel-title pull-left">Gene set enrichment</h3>
	    <div class="btn-group pull-right">
	      <a class="btn btn-default btn-sm" data-toggle="collapse" href="#gseaDesc">
		<i class="fa fa-lg fa-question"></i>
	      </a>
	      <a class="btn btn-default btn-sm" data-toggle="collapse" href="#gseaOptions" aria-expanded="false" aria-controls="gseaOptions">
		<i class="fa fa-gear"></i>
	      </a>
	      <a class="btn btn-default btn-sm" data-toggle="collapse" href="#gseaClose">
		<i class="fa fa-lg fa-minus-square-o"></i>
	      </a>
	    </div>
	    <div class="clearfix"></div>
	    <div class="collapse" id="gseaDesc">
	      <div class="card card-block">
		Gene set enrichment testing by hyper-geometric test. 
	      </div>
	    </div>	  
	    <div class="collapse" id="gseaOptions">
	      Gene set:
	      <select id="geneset_selection" class="btn btn-default">
	      
	      </select>
	      <button type="button" id="rerunGSEA" class="btn btn-primary btn-block">retest gene set</button>
	    </div>
	  </div>
	  <div class="panel-body collapse in" id="gseaClose">
	    <div style="height:330px" id="gsea_panel">
	      <div id="mhg-chart"></div>
	    </div>
	  </div>
	</div>
      </div>     

  </div>

  <!--- gene row --->
  <div class="row"><div class="col-sm-12">
    <div class="panel panel-default">
      <div class="panel-heading">
	<h3 class="panel-title pull-left">Gene expression</h3>
	<div class="btn-group pull-right">
	  <a class="btn btn-default btn-sm" data-toggle="collapse" href="#gexpDesc">
	    <i class="fa fa-lg fa-question"></i>
	  </a>
	  <a class="btn btn-default btn-sm" data-toggle="collapse" href="#gexpOptions" aria-expanded="false" aria-controls="gexpOptions">
	    <i class="fa fa-gear"></i>
	  </a>
	  <a class="btn btn-default btn-sm" data-toggle="collapse" href="#gexpClose" aria-expanded="false" aria-controls="pcaClose">
	    <i class="fa fa-lg fa-minus-square-o"></i>
	  </a>
	</div>
	<div class="clearfix"></div>
	<div class="collapse" id="gexpDesc">
	  <div class="card card-block">
	    Plot overall distribution of gene expression or compare between 2 main groups. Use options to select mode of visualization. Additional gene information is also provided. 
	  </div>
	</div>	  
	<div class="collapse" id="gexpOptions">
	  Gene:
	  <select id="gexp_gene_selection" class="btn btn-default">
	  </select>	    
	  Plot type:
	  <select id="gexp_plottype" class="btn btn-default">
	    <option value="box_plot" selected>Box Plot</option>
	    <option value="notched_box_plot">Notched Box Plot</option>
	    <option value="violin_plot">Violin Plot</option>
	      <option value="bean_plot">Bean Plot</option>
	    <option value="beeswam_plot">Beeswam Plot</option>
	    <option value="scatter_plot">Scatterplot Plot</option>
	  </select>
	    <button type="button" id="replotGexp" class="btn btn-primary btn-block">replot gene expression</button>
	</div>	    
      </div>
      <div class="panel-body collapse in" id="gexpClose">
	<div class="row">
	  <div class="col-sm-3">
	    <div style="height:336px" id="gexp_panel1">
	      <b>Overall</b>
	      <div class="chart-wrapper" id="gexp-chart-distro1"></div>
	    </div>
	  </div>
	  <div class="col-sm-4">
	    <div style="height:336px" id="gexp_panel2">
	      <b>By Group</b>
	      <div class="chart-wrapper" id="gexp-chart-distro2"></div>
	    </div>
	  </div>
	  <div class="col-sm-5">
	    <div style="height:336px; overflow-y:scroll;" id="gexp_info">
	      <b>Information</b>
	      <div id="savedResults"></div>
	      <div class="clearfix"></div>
	    </div>
	  </div>
	</div>
      </div>
    </div>
  </div></div>
  
</div>

<script src="js/clusterfck.js"></script>
<script src="js/clusterfck_custom.js"></script>
<script src="js/mannwhitneyu.js"></script>
<script src="js/pca.js"></script>
<script src="js/column-chart.js"></script>

<script src="js/mhg.js"></script>
<script src="js/mhg_custom.js"></script>

<script src="js/qc.js"></script>
<script src="js/pcscatter.js"></script>
<script src="js/heatmap.js"></script>
<script src="js/ubit2.js"></script>
<script src="js/org.Hs.GO2Symbol.js"></script> 

<script src="js/gexp.js"></script> 
<script src="js/distrochart.js"></script> 

<script>
  var dataInit = [
  ["","KIT","PROM1","CXCR4","PARP2","FOXA1","CDH1","EMP1","ERBB3","MUC1","CD24","ERBB2","PTEN","AR","KRT19","GATA3","CAV2","MYLK","PDGFRA","KRT5","CAV1","JAG1","COL1A2","PLCB4","MTOR","NOTCH3","BCL2L1","ACTA2","LGR5","MYCN","CCNB1"],
  ["SAMPLE.1 Individual #1 Basal",999,26.35196313,999,999,999,27.08864793,29.76834572,999,999,999,25.99769589,24.90127283,999,23.66209212,999,22.70504017,22.4103863,25.11302594,19.8101043,24.47289556,22.21428545,25.53932598,21.75975964,999,26.44337452,18.19240437,25.18622502,25.76831471,24.42517983,25.70503837],
  ["SAMPLE.2 Individual #1 Basal",999,23.12723537,999,25.83642447,999,23.71587187,999,999,26.61100016,999,999,24.25037013,999,23.03658626,999,22.74139312,26.35614184,31.40060091,19.27909769,24.67709034,23.78017502,999,22.12476627,999,33.96291481,18.86592865,999,21.5060959,999,999],
  ["SAMPLE.3 Individual #1 Basal",999,24.78442987,999,999,999,999,27.03926534,26.95237058,999,999,999,27.16599711,999,26.28752055,999,25.80771878,24.78627491,25.48710427,21.4438051,25.18813133,23.75668627,26.94801755,28.88069564,999,33.64691783,18.7324904,23.00475749,21.96971005,22.79206965,28.07387452],
  ["SAMPLE.4 Individual #1 Basal",25.94803898,24.27772022,999,999,999,27.90792038,26.64938389,999,999,999,999,999,999,22.69715558,25.81631557,25.62873852,24.26348133,25.9291376,20.24535794,26.31174867,23.91467143,27.7434859,22.86149617,999,30.34264314,17.60272944,27.4407271,22.44331127,999,999],
  ["SAMPLE.5 Individual #1 Basal",999,22.81000433,999,999,999,23.88691966,999,28.03046271,24.89900501,999,33.97732956,999,25.46060574,22.68672481,25.35435262,22.16931541,25.43300141,24.79856893,19.63637175,26.93243894,23.50235813,25.1049184,23.07719659,999,25.93476534,19.09332212,24.5720493,22.49044559,999,999],
  ["SAMPLE.6 Individual #1 Basal",999,24.73899247,999,999,999,999,26.40887581,27.12640628,999,999,999,24.54401452,999,25.2040805,999,22.45241313,22.96488968,24.99314787,20.59781362,23.21541769,23.6595374,25.12099184,22.38141133,999,24.6357447,18.58079196,999,23.34614288,32.93778104,25.71955514],
  ["SAMPLE.7 Individual #1 Basal",999,22.85394997,999,999,999,999,999,999,999,999,22.71904758,999,27.16518579,22.94382047,999,24.65333527,24.81997486,27.73704022,19.9627355,999,22.75734589,25.78451013,20.29794549,999,29.68990882,17.54433013,25.83550662,22.09553123,25.3524042,25.70677637],
  ["SAMPLE.8 Individual #1 Basal",999,29.12682674,999,999,999,999,25.30776373,999,999,999,33.50279032,28.48754602,999,23.62817235,999,23.10023679,22.66434314,26.72458057,19.73462814,26.31070013,22.42494855,28.08598559,23.45043174,999,24.8188849,19.37253092,27.66541534,22.33141804,999,999],
  ["SAMPLE.9 Individual #1 Basal",999,999,999,999,999,25.76117242,999,33.75249588,999,999,27.64053197,999,999,21.11849653,25.59498779,20.35313017,25.16949254,23.74671309,19.29219263,24.71897522,23.10358884,24.27972535,21.68326819,999,24.79746908,17.6610602,23.21734856,22.97442211,33.17310503,999],
  ["SAMPLE.10 Individual #1 Basal",999,23.45497736,999,999,999,999,28.51704147,999,23.10701539,999,25.78558313,25.12362634,999,24.05694338,26.68948118,22.54283081,24.98903674,24.25668503,19.76101829,25.25167364,25.41292758,25.92511155,23.3703414,999,29.49438424,19.15163619,25.55490577,24.1067025,999,999],
  ["SAMPLE.11 Individual #1 Basal",999,23.72760324,999,999,999,999,27.12963023,999,999,999,25.965308,25.40673915,26.92725351,23.38786816,999,22.37757168,22.63589678,23.66984638,19.88622269,25.85066712,23.739477,24.55263198,24.86404662,999,31.54469934,18.39958595,999,22.00124087,23.3663071,27.53031882],
  ["SAMPLE.12 Individual #1 Basal",27.3558229,999,999,999,999,24.98702588,999,999,25.49182689,999,22.10741344,999,999,21.77878522,999,21.41646848,23.80203984,24.26947181,20.17495923,33.93366593,21.53159603,25.00153379,23.02337135,999,25.6886113,18.23848901,25.86291963,999,999,999],
  ["SAMPLE.13 Individual #1 Basal",999,27.77718656,999,999,999,999,27.16905584,999,26.26501562,999,23.88270611,26.83996979,23.86551191,23.72534984,999,23.8050678,23.05365107,23.72859839,19.40495802,28.58520131,23.51343871,24.70033151,21.55660628,999,28.24889452,17.72789829,999,22.12433913,20.91430349,999],
  ["SAMPLE.14 Individual #1 Basal",999,24.96932729,999,27.63685176,999,999,26.77793388,25.62159416,29.11925422,999,26.10400172,23.95670861,25.62339893,22.83816972,999,25.51309986,24.53220691,23.22962328,19.33098545,999,23.15839196,23.19824376,999,999,31.46963231,17.60850236,999,21.5585151,22.87298881,25.55933108],
  ["SAMPLE.15 Individual #1 Basal",999,27.00650828,999,999,999,31.08524099,999,999,999,999,25.68224782,25.48809015,999,22.49978638,999,22.69557456,26.60009046,24.96933555,20.6974307,23.52409436,25.23969041,25.7275211,21.90015286,999,31.09131228,18.57456233,27.39330554,25.10934891,27.92701814,999],
  ["SAMPLE.16 Individual #1 Basal",33.49749521,24.27772968,999,31.00073657,999,999,999,999,999,999,999,24.55229951,999,23.7223756,26.46169794,23.27711943,999,23.33764594,20.5653628,24.2978356,24.9379165,23.95257419,21.63269084,999,31.3266691,18.24328149,999,26.36060951,22.16833471,999],
  ["SAMPLE.17 Individual #1 Basal",33.20480443,25.69123901,999,33.7114051,999,26.65452119,999,999,23.24847812,999,25.73038673,24.52500447,999,22.35235825,999,24.51571099,999,24.94138213,20.12524026,24.28788485,23.75407969,24.29737522,999,999,31.41958502,18.52837255,27.38255008,21.97291518,23.02708534,23.01790487],
  ["SAMPLE.18 Individual #1 Basal",999,999,999,28.10098077,999,999,999,999,999,999,33.15037219,29.93837208,999,23.94368528,999,23.31739529,22.07252172,28.2291677,19.93641883,24.93156731,22.87614119,999,22.19046187,999,26.17758331,17.76547322,25.87378067,26.17103054,23.47654141,999],
  ["SAMPLE.19 Individual #1 Basal",999,25.08316708,999,26.72390376,999,999,25.75411607,999,24.96826402,999,22.41404653,24.89313194,999,23.45965915,999,23.8810004,25.29162144,25.35462019,20.32930699,26.23956945,24.2304622,26.42981344,20.77858735,999,23.0005492,18.63722305,28.22282922,22.45496851,23.59815688,24.97318109],
  ["SAMPLE.20 Individual #1 Basal",999,24.9522851,999,999,999,25.26787684,999,999,26.27625309,999,30.59332539,24.54097015,999,24.43044215,999,22.63108524,999,999,21.10448007,25.68064273,22.87338961,999,22.26381731,999,24.75376607,17.76125594,27.36454343,26.70052968,999,999],
  ["SAMPLE.50 Individual #1 Luminal",999,28.20677041,999,999,999,24.34907659,24.90887334,26.9325567,19.97249564,24.80509134,23.96232054,24.02404194,25.32833589,24.51553325,999,999,999,33.02707772,999,26.30728572,23.65220823,999,999,999,999,20.98329512,999,22.57307944,999,999],
  ["SAMPLE.51 Individual #1 Luminal",999,22.94901298,999,999,999,24.23295265,21.61854175,24.06975443,19.66795448,999,33.24381644,23.74933001,27.08495904,22.46435218,24.86571818,999,999,29.63569739,26.75199275,28.68090325,21.73318924,32.38137328,999,999,31.90947764,21.80521881,999,29.8717194,32.72052508,999],
  ["SAMPLE.52 Individual #1 Luminal",999,23.2176697,999,25.04471028,999,26.07783332,26.68439525,26.17583025,19.70016561,25.10137977,25.74248857,24.5394077,25.26219801,22.73535824,24.52083432,999,999,999,999,25.65233159,999,999,22.08602894,999,999,25.55911139,999,999,999,999],
  ["SAMPLE.53 Individual #1 Luminal",999,28.20677156,999,27.02096849,999,24.31152876,25.33269249,25.90604519,19.79290433,24.25277292,24.50666214,26.45610884,24.27638911,25.4259152,24.04002254,999,999,30.56327116,999,999,22.86502842,999,999,999,999,22.90838134,999,23.98332973,999,999],
  ["SAMPLE.54 Individual #1 Luminal",999,23.55242173,999,999,999,23.40837991,21.89680136,25.81714554,19.9367677,23.33233777,25.60251319,23.55238352,24.73245242,21.76897198,25.09369015,22.75571231,999,999,999,23.58077749,22.86875555,999,999,999,999,23.44981935,999,999,999,999],
  ["SAMPLE.55 Individual #1 Luminal",999,999,999,23.43901903,999,22.58990689,23.42396283,25.60179232,19.71623149,23.9440805,22.51512332,22.59500596,24.89440192,21.01155888,24.92771969,999,999,32.44051161,28.99596623,999,999,33.57826495,999,999,25.26121973,31.06513925,999,29.6467734,27.41695323,999],
  ["SAMPLE.56 Individual #1 Luminal",999,25.34772314,999,24.09836974,999,24.61935793,24.75925825,999,18.75187506,24.21402325,25.53851048,25.35790946,26.55164389,26.00257588,999,999,26.4305425,31.07584262,22.07917239,27.14046274,999,999,25.5668995,999,999,21.58583986,999,32.05553937,999,999],
  ["SAMPLE.57 Individual #1 Luminal",999,23.12305995,999,27.10326662,999,26.29675794,26.35788392,27.93766568,19.8518875,26.21630863,26.4109633,23.65607477,26.08963339,23.05655524,23.34847624,24.98684458,999,25.43422399,23.97009524,26.93086187,999,26.34086433,999,999,999,31.26407986,999,999,22.95790122,999],
  ["SAMPLE.58 Individual #1 Luminal",999,999,999,25.55750054,999,22.932807,23.25166952,27.15246681,999,24.97115261,999,24.14990256,27.45813077,23.21249678,25.81840575,999,999,23.27165284,23.18258284,27.82720395,25.40170682,24.50965869,999,999,999,26.40414247,999,999,23.27366192,999],
  ["SAMPLE.59 Individual #1 Luminal",999,999,999,24.09656219,999,23.78541283,22.56414854,25.80699794,20.62534366,23.79296767,26.09122116,23.98312746,25.10180031,23.59598241,27.04219806,23.46278382,999,999,23.84639908,999,23.90348977,999,24.8207731,999,999,23.13678789,999,28.16522379,999,999],
  ["SAMPLE.60 Individual #1 Luminal",25.39425799,21.5201385,999,999,999,999,22.63362977,999,23.21669738,24.20447618,21.94028818,999,25.52905314,22.62743364,999,999,999,999,24.40370898,999,999,999,999,999,25.01502463,26.55782576,999,21.47990137,27.76347619,999],
  ["SAMPLE.61 Individual #1 Luminal",999,999,999,999,999,23.74055508,25.38550588,24.73405772,21.69292161,25.17970753,26.2210955,24.03863223,25.35701627,23.26037225,25.71480707,999,999,999,999,999,999,999,999,999,999,25.40254156,999,32.57682251,999,999],
  ["SAMPLE.62 Individual #1 Luminal progenitor",22.60808882,23.30469868,999,23.67220238,999,26.69328645,25.1348303,25.63218023,26.14653185,23.86669773,999,24.35460211,999,23.90786941,25.74007946,23.51995676,999,25.50220529,999,999,24.88093508,26.12562967,26.41645895,999,999,23.71386752,999,32.80656866,999,999],
  ["SAMPLE.63 Individual #1 Luminal progenitor",24.18450386,24.65173782,999,25.34540193,999,23.20099731,22.88231156,999,21.72644597,24.14212268,999,23.53049601,999,23.37442565,25.24993779,21.73160548,999,999,21.96017493,999,23.3936471,999,999,999,24.83062438,32.31587411,999,999,999,999],
  ["SAMPLE.64 Individual #1 Luminal progenitor",999,25.81267844,999,999,999,25.04296193,26.68762557,27.09280267,24.33365794,999,999,24.63867309,27.53400368,32.77469614,999,999,999,999,23.75278729,25.07052792,25.58488877,999,999,999,25.59542635,21.36106941,999,26.1542471,999,25.29768754],
  ["SAMPLE.65 Individual #1 Luminal progenitor",999,23.99819595,999,25.9681144,999,25.57221509,26.08957719,27.11423367,999,25.40744485,29.70377952,26.37357832,24.90743664,24.31584851,999,999,999,999,26.4172189,26.98962795,23.81157349,999,999,999,999,23.0541023,999,31.62545608,999,999],
  ["SAMPLE.66 Individual #1 Luminal progenitor",22.23623209,23.12634952,999,24.74032696,999,26.59351807,22.89681766,32.57300481,22.41442958,24.82727534,29.84790031,22.92377008,24.60637282,23.85470023,25.51074729,24.35613721,999,999,26.12216028,999,24.60588222,999,999,999,999,22.44917863,999,999,999,999],
  ["SAMPLE.67 Individual #1 Luminal progenitor",999,25.49488239,999,24.29987516,999,999,27.62005216,26.87686674,24.44809823,999,999,27.4124126,999,25.19409691,26.8778656,999,999,24.54731155,23.72840117,24.77841834,999,25.68910309,999,999,26.32632303,999,999,999,999,999],
  ["SAMPLE.68 Individual #1 Luminal progenitor",999,24.41634338,999,999,999,999,23.39041007,27.12107571,20.72009953,24.73168085,999,22.50996543,999,22.81980348,999,22.6355848,999,999,999,999,999,999,29.46302196,999,999,22.96323004,999,28.81628329,999,28.40305517],
  ["SAMPLE.69 Individual #1 Luminal progenitor",22.78915128,23.62957538,999,23.12605398,999,26.55676987,999,26.86069587,22.38852727,25.52432812,999,23.66905837,999,22.54384077,26.10471456,999,999,999,22.65920612,999,999,999,999,999,23.84184321,22.30857817,999,32.39727605,999,999],
  ["SAMPLE.70 Individual #1 Luminal progenitor",26.30961734,23.52141837,999,23.22510962,999,24.33072915,23.86213611,27.258637,999,24.03394353,27.07418904,22.88695628,26.21040157,25.36632112,999,23.07920207,999,999,25.74145042,23.98847943,23.31429388,999,25.92881693,999,25.01372889,21.15814286,999,24.27876843,999,999]
  ];

  var container = document.getElementById('incsv');
  var hot = new Handsontable(container, {
  data: dataInit,
  rowHeaders: false,
  colHeaders: false
  });

</script>   

<script type="text/javascript">  
  function draw(){
    initQC();
    initDiffexpPca();
    initHeatmap();    
    drawMhg();
    initGexp(); // has to be after initDiffexpPca since uses ordering
    drawGexp();
  };

  function init() {
    initData(); 
    initMhg();
    draw();
  };

  $(window).load(init);
  $("#inbut").click(init);
  $(window).resize(draw);

  $("#rerunDiffexp").click(function() {initDiffexpPca(); initGexp(); drawMhg(); initHeatmap();});
  $("#rerunKmeans").click(initDiffexpPca);
  $("#rerunGSEA").click(drawMhg);
  
  $("#rerunHclust").click(initHeatmap);
  $("#rerun_percentage_success_per_gene").click(init);
  $("#rerun_percentage_genes_detected_per_sample").click(init);
  $("#replotGexp").click(drawGexp);

  $(function () {
  $("#sortable").sortable();
  $("#sortable").disableSelection();
  });
</script>

