
function Help(id)
{    
    if (id == 0) alert("Here you can load a curated genome model (homo sapiens or mus musculus). These genome models will be updated on a regular basis according to the newest genome and GENCODE releases."); //existing gene model
    if (id == 1) alert("Specify the chromosomes you want to include in the genome model. Consult the UCSC website (http://hgdownload.cse.ucsc.edu/downloads.html) for the specific chromosome nomenclature in 2bit files. Of note, the order of the chromosomes entered dictates the order of the chromosomes in the genome model that will be built by BrowserGenome. Enter chromosomes separated by commas without spaces."); //Choose chomosomes to be included
    if (id == 2) alert("Download the genome sequence file of interest from the UCSC website (http://hgdownload.cse.ucsc.edu/downloads.html)."); //Load genome sequence file (.2bit format)
    if (id == 3) alert("Download the gene model file of interest from the GENCODE website (http://www.gencodegenes.org/)."); //Load gene model (GENCODE .gtf format, can take several minutes)
    if (id == 4) alert("A custom genome model file will typically encompass 3-4 MB on your local hard drive. Save this model for later use (see menu item 3 in menu 'Choose existing genome model')."); //Save the genome model in a BrowserGenome-specific format (.BrowGenModel)
    if (id == 5) alert("Download the genome sequence file from the UCSC website (http://hgdownload.cse.ucsc.edu/downloads.html)."); //Load genome sequence hg38.2bit
    if (id == 6) alert("In case the sequencing reads in the FASTQ file specified encompass several multiplexed samples, enter here the 0-based position and the sequence of the barcode that marks read sequences to be filtered out for analysis."); //What is the position and sequence of an internal barcode?
    if (id == 7) alert("Some RNA sequencing libraries do not contain the mappable sequence at the start of the sequencing reads, e.g. because the library molecules contain a constant sequence for PCR amplification or a barcode sequence next to the sequencing primer binding site. In this case, this option allows to specify a 0-based position within sequencing reads at which BrowserGenome should start read mapping."); //At what position in your reads should the 25-mer alignment start?
    if (id == 8) alert("In case the first 25-mer fails to perfectly map to the genome, which can be due to a polymorphism, an exon-exon junction, or a sequencing error, this option allows BrowserGenome to try to map a second downstream 25-mer.The chance that also the second 25-mer fails to map to the genome due to said reasons is small."); //If an alignment fails: Allow alignement of a downstream 25-mer?
    if (id == 9) alert("On computers with 32 GB RAM or more, choosing the high memory mode speeds up the mapping process roughly ten-fold."); //How much memory can BrowserGenome use?
    if (id == 10) alert("The mapping process can be stopped at any time by clicking STOP. In that case, the fraction of reads that have been mapped so far is still available for evaluation. Typically, 18 mio. reads will be mapped per hour."); //Start the mapping
    if (id == 11) alert("Clicking EXPORT will create a tab-delimited text file listing gene symbols and corresponding mapped read numbers, optionally normalized. This file can be opened e.g. with Microsoft Excel."); //Export the data 
    if (id == 12) alert("Only exact matches will be found."); //Rapid sequence search 
}
